web/source/text/defaultOutput.ts

```diff
@@ -211,30 +211,34 @@ namespace com.keyman.text {
 
     // Test for otherwise unimplemented keys on the the base default & shift layers.
     // Those keys must be blocked by keyboard rules if intentionally unimplemented; otherwise, this function will trigger.
-    public static forBaseKeys(Lkc: KeyEvent, keyShiftState: number, ruleBehavior?: RuleBehavior) {
+    public static forBaseKeys(Lkc: KeyEvent, ruleBehavior?: RuleBehavior) {
       let n = Lkc.Lcode;
+      let keyShiftState = Lkc.Lmodifiers;
+
       // check if exact match to SHIFT's code.  Only the 'default' and 'shift' layers should have default key outputs.
-      if(keyShiftState != 0 && keyShiftState != 1) {
+      if(keyShiftState == Codes.modifierCodes['SHIFT']) {
+        keyShiftState = 1;
+      } else if(keyShiftState != 0) {
         if(ruleBehavior) {
           ruleBehavior.warningLog = "KMW only defines default key output for the 'default' and 'shift' layers!";
+          return '';
         }
       }
 
-      if(keyShiftState == 0 || keyShiftState == 1) { // keyShiftState can only be 0 or 1.  (1 being mapped from Codes.modifier['SHIFT'] by the caller.)
-        try {
-          if(n >= Codes.keyCodes['K_0'] && n <= Codes.keyCodes['K_9']) { // The number keys.
-            return Codes.codesUS[keyShiftState][0][n-Codes.keyCodes['K_0']];
-          } else if(n >= Codes.keyCodes['K_A'] && n <= Codes.keyCodes['K_Z']) { // The base letter keys
-            return String.fromCharCode(n+(keyShiftState?0:32));  // 32 is the offset from uppercase to lowercase.
-          } else if(n >= Codes.keyCodes['K_COLON'] && n <= Codes.keyCodes['K_BKQUOTE']) {
-            return Codes.codesUS[keyShiftState][1][n-Codes.keyCodes['K_COLON']];
-          } else if(n >= Codes.keyCodes['K_LBRKT'] && n <= Codes.keyCodes['K_QUOTE']) {
-            return Codes.codesUS[keyShiftState][2][n-Codes.keyCodes['K_LBRKT']];
-          }
-        } catch (e) {
-          if(ruleBehavior) {
-            ruleBehavior.errorLog = "Error detected with default mapping for key:  code = " + n + ", shift state = " + (keyShiftState == 1 ? 'shift' : 'default');
-          }
+      // Now that keyShiftState is either 0 or 1, we can use the following structure to determine the default output.
```

Comment:
If `ruleBehavior` is not defined, `keyShiftState` could be non-zero by the code above (no `return ''` in that situation).